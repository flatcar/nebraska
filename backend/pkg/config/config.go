package config

import (
	"errors"
	"flag"
	"fmt"
	"net/url"
	"os"

	"github.com/knadh/koanf"
	"github.com/knadh/koanf/providers/basicflag"

	"github.com/flatcar/nebraska/backend/pkg/random"
)

type Config struct {
	EnableSyncer        bool   `koanf:"enable-syncer"`
	HostFlatcarPackages bool   `koanf:"host-flatcar-packages"`
	FlatcarPackagesPath string `koanf:"flatcar-packages-path"`
	NebraskaURL         string `koanf:"nebraska-url"`
	SyncerPkgsURL       string `koanf:"syncer-packages-url"`
	HTTPLog             bool   `koanf:"http-log"`
	HTTPStaticDir       string `koanf:"http-static-dir"`
	AuthMode            string `koanf:"auth-mode"`
	FlatcarUpdatesURL   string `koanf:"sync-update-url"`
	CheckFrequencyVal   string `koanf:"sync-interval"`
	AppLogoPath         string `koanf:"client-logo"`
	AppTitle            string `koanf:"client-title"`
	AppHeaderStyle      string `koanf:"client-header-style"`
	APIEndpointSuffix   string `koanf:"api-endpoint-suffix"`
	Debug               bool   `koanf:"debug"`
	ServerPort          uint   `koanf:"port"`
	RollbackDBTo        string `koanf:"rollback-db-to"`

	GhClientID        string `koanf:"gh-client-id"`
	GhClientSecret    string `koanf:"gh-client-secret"`
	GhSessionAuthKey  string `koanf:"gh-session-secret"`
	GhSessionCryptKey string `koanf:"gh-session-crypt-key"`
	GhWebhookSecret   string `koanf:"gh-webhook-secret"`
	GhReadWriteTeams  string `koanf:"gh-rw-teams"`
	GhReadOnlyTeams   string `koanf:"gh-ro-teams"`
	GhEnterpriseURL   string `koanf:"gh-enterprise-url"`

	OidcClientID      string `koanf:"oidc-client-id"`
	OidcIssuerURL     string `koanf:"oidc-issuer-url"`
	OidcAdminRoles    string `koanf:"oidc-admin-roles"`
	OidcViewerRoles   string `koanf:"oidc-viewer-roles"`
	OidcRolesPath     string `koanf:"oidc-roles-path"`
	OidcScopes        string `koanf:"oidc-scopes"`
	OidcManagementURL string `koanf:"oidc-management-url"`
	OidcLogoutURL     string `koanf:"oidc-logout-url"`
	OidcAudience      string `koanf:"oidc-audience"`
}

const (
	oidcClientIDEnvName      = "NEBRASKA_OIDC_CLIENT_ID"
	ghClientIDEnvName        = "NEBRASKA_GITHUB_OAUTH_CLIENT_ID"
	ghClientSecretEnvName    = "NEBRASKA_GITHUB_OAUTH_CLIENT_SECRET"
	ghSessionAuthKeyEnvName  = "NEBRASKA_GITHUB_SESSION_SECRET"
	ghSessionCryptKeyEnvName = "NEBRASKA_GITHUB_SESSION_CRYPT_KEY"
	ghWebhookSecretEnvName   = "NEBRASKA_GITHUB_WEBHOOK_SECRET"
	ghEnterpriseURLEnvName   = "NEBRASKA_GITHUB_ENTERPRISE_URL"
)

func (c *Config) Validate() error {
	if c.HostFlatcarPackages {
		if c.FlatcarPackagesPath == "" {
			return errors.New("invalid Flatcar packages path. Please ensure you provide a valid path using -flatcar-packages-path")
		}

		tmpFile, err := os.CreateTemp(c.FlatcarPackagesPath, "")
		if err != nil {
			return fmt.Errorf("invalid Flatcar packages path: %w", err)
		}
		defer os.Remove(tmpFile.Name())

		if _, err := url.ParseRequestURI(c.NebraskaURL); err != nil {
			return errors.New("invalid Nebraska URL, please ensure the value provided using -nebraska-url is a valid url")
		}
	}

	switch c.AuthMode {
	case "github":
		if c.GhClientID == "" || c.GhClientSecret == "" || c.GhReadOnlyTeams == "" || c.GhReadWriteTeams == "" {
			return errors.New("invalid github configuration")
		}
	case "oidc":
		if c.OidcClientID == "" || c.OidcIssuerURL == "" || c.OidcAdminRoles == "" || c.OidcViewerRoles == "" {
			return errors.New("invalid OIDC configuration")
		}
	}

	return nil
}

func Parse() (*Config, error) {
	var config Config

	f := flag.NewFlagSet("config", flag.ContinueOnError)
	f.String("rollback-db-to", "", "Rollback db migration to the provided id, eg:0003")
	f.Bool("enable-syncer", false, "Enable Flatcar packages syncer")
	f.Bool("host-flatcar-packages", false, "Host Flatcar packages in Nebraska")
	f.String("flatcar-packages-path", "", "Path where Flatcar packages files should be stored")
	f.String("nebraska-url", "http://localhost:8000", "nebraska URL (http://host:port - required when hosting Flatcar packages in nebraska)")
	f.String("syncer-packages-url", "", "use this URL instead of the original one for packages created by the syncer; any {{ARCH}} and {{VERSION}} in the URL will be replaced by the original package's architecture and version, respectively. If this option is not used but the 'host-flatcar-packages' one is, then the URL will be nebraska-url/flatcar/ .")
	f.Bool("http-log", false, "Enable http requests logging")
	f.String("http-static-dir", "../frontend/build", "Path to frontend static files")
	f.String("auth-mode", "oidc", "authentication mode, available modes: noop, github, oidc")

	f.String("gh-client-id", "", fmt.Sprintf("GitHub client ID used for authentication; can be taken from %s env var too", ghClientIDEnvName))
	f.String("gh-client-secret", "", fmt.Sprintf("GitHub client secret used for authentication; can be taken from %s env var too", ghClientSecretEnvName))
	f.String("gh-session-secret", "", fmt.Sprintf("Session secret used for authenticating sessions in cookies used for storing GitHub info , will be generated if none is passed; can be taken from %s env var too", ghSessionAuthKeyEnvName))
	f.String("gh-session-crypt-key", "", fmt.Sprintf("Session key used for encrypting sessions in cookies used for storing GitHub info, will be generated if none is passed; can be taken from %s env var too", ghSessionCryptKeyEnvName))
	f.String("gh-webhook-secret", "", fmt.Sprintf("GitHub webhook secret used for validing webhook messages; can be taken from %s env var too", ghWebhookSecretEnvName))
	f.String("gh-rw-teams", "", "comma-separated list of read-write GitHub teams in the org/team format")
	f.String("gh-ro-teams", "", "comma-separated list of read-only GitHub teams in the org/team format")
	f.String("gh-enterprise-url", "", fmt.Sprintf("base URL of the enterprise instance if using GHE; can be taken from %s env var too", ghEnterpriseURLEnvName))

	f.String("oidc-client-id", "", fmt.Sprintf("OIDC client ID used for authentication;can be taken from %s env var too", oidcClientIDEnvName))
	f.String("oidc-issuer-url", "", "OIDC issuer URL used for authentication")
	f.String("oidc-admin-roles", "", "comma-separated list of accepted roles with admin access")
	f.String("oidc-viewer-roles", "", "comma-separated list of accepted roles with viewer access")
	f.String("oidc-roles-path", "roles", "json path in which the roles array is present in the id token")
	f.String("oidc-scopes", "openid,profile,email", "comma-separated list of scopes to be used in OIDC")
	f.String("oidc-management-url", "", "OIDC management url for managing the account")
	f.String("oidc-logout-url", "", "OIDC logout URL (optional fallback when end_session_endpoint is not available in discovery)")
	f.String("oidc-audience", "", "OIDC audience parameter for the access token")
	f.String("sync-update-url", "https://public.update.flatcar-linux.net/v1/update/", "Flatcar update URL to sync from")
	f.String("sync-interval", "1h", "Sync check interval (the minimum depends on the number of channels to sync, e.g., 8m for 8 channels incl. different architectures)")
	f.String("client-logo", "", "Client app logo, should be a path to svg file")
	f.String("client-title", "", "Client app title")
	f.String("client-header-style", "light", "Client app header style, should be either dark or light")
	f.String("api-endpoint-suffix", "", "Additional suffix for the API endpoint to serve Omaha clients on; use a secret to only serve your clients, e.g., mysecret results in /v1/update/mysecret")
	f.Bool("debug", false, "sets log level to debug")
	f.Uint("port", 8000, "port to run server")

	k := koanf.New(".")

	// Load from flag if args are provided
	if len(os.Args) == 0 {
		return nil, errors.New("no args provided")
	}

	if err := f.Parse(os.Args[1:]); err != nil {
		return nil, fmt.Errorf("error parsing flags: %w", err)
	}
	if err := k.Load(basicflag.Provider(f, "."), nil); err != nil {
		return nil, fmt.Errorf("error loading config from flags: %w", err)
	}

	if err := k.Unmarshal("", &config); err != nil {
		return nil, fmt.Errorf("error unmarshal config: %w", err)
	}

	switch config.AuthMode {
	case "oidc":
		config.OidcClientID = getPotentialOrEnv(config.OidcClientID, oidcClientIDEnvName)

	case "github":
		config.GhClientID = getPotentialOrEnv(config.GhClientID, ghClientIDEnvName)
		config.GhClientSecret = getPotentialOrEnv(config.GhClientSecret, ghClientSecretEnvName)
		config.GhSessionAuthKey = getPotentialOrEnv(config.GhSessionAuthKey, ghSessionAuthKeyEnvName)
		config.GhSessionCryptKey = getPotentialOrEnv(config.GhSessionCryptKey, ghSessionCryptKeyEnvName)
		config.GhWebhookSecret = getPotentialOrEnv(config.GhWebhookSecret, ghWebhookSecretEnvName)
		config.GhEnterpriseURL = getPotentialOrEnv(config.GhEnterpriseURL, ghEnterpriseURLEnvName)

		if config.GhSessionAuthKey == "" {
			config.GhSessionAuthKey = string(random.Data(32))
		}

		if config.GhSessionCryptKey == "" {
			config.GhSessionCryptKey = string(random.Data(32))
		}
	}

	return &config, nil
}

func getPotentialOrEnv(potentialValue, envName string) string {
	if potentialValue != "" {
		return potentialValue
	}
	return os.Getenv(envName)
}
