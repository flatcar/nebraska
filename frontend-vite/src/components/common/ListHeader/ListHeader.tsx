import { Box } from '@mui/material';
import Grid from '@mui/material/Grid';
import Typography from '@mui/material/Typography';
import makeStyles from '@mui/styles/makeStyles';
import React from 'react';

const useStyles = makeStyles({
  sectionHeader: {
    padding: '1em',
  },
});

export default function ListHeader(props: { title: string; actions?: React.ReactElement[] }) {
  const classes = useStyles();
  const actions = props.actions || [];
  return (
    <Grid
      container
      alignItems="flex-start"
      justifyContent="space-between"
      className={classes.sectionHeader}
    >
      {props.title && (
        <Grid item>
          <Box p={0.5}>
            <Typography variant="h1">{props.title}</Typography>
          </Box>
        </Grid>
      )}
      {actions &&
        actions.map((action, i: number) => (
          <Grid item key={i}>
            {action}
          </Grid>
        ))}
    </Grid>
  );
}
