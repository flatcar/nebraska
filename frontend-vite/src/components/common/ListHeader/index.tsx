export { default } from './ListHeader';
export * from './ListHeader';
