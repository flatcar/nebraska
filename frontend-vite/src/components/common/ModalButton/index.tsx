export { default } from './ModalButton';
export * from './ModalButton';
