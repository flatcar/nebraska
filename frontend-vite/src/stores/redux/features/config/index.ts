import { createSlice, PayloadAction } from '@reduxjs/toolkit';

export const CONFIG_STORAGE_KEY = 'nebraska_config';

export interface NebraskaConfig {
  title: string;
  nebraska_config: string;
  nebraska_version: string;
  login_url: string;
  auth_mode: string;
  header_style: string;
  [prop: string]: any;
}

const nebraskaConfig = localStorage.getItem(CONFIG_STORAGE_KEY) || '{}';
const initialState: NebraskaConfig = JSON.parse(nebraskaConfig) || ({} as NebraskaConfig);

export const configSlice = createSlice({
  name: 'config',
  initialState,
  reducers: {
    setConfig: (state, action: PayloadAction<NebraskaConfig>) => {
      Object.assign(state, action.payload);
      localStorage.setItem(CONFIG_STORAGE_KEY, JSON.stringify(state));
    },
  },
});

export const { setConfig } = configSlice.actions;
export default configSlice.reducer;
