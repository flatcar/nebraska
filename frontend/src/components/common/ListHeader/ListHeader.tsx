import { Box } from '@mui/material';
import Grid from '@mui/material/Grid';
import { styled } from '@mui/material/styles';
import Typography from '@mui/material/Typography';
import React from 'react';

const PREFIX = 'ListHeader';

const classes = {
  sectionHeader: `${PREFIX}-sectionHeader`,
};

const StyledGrid = styled(Grid)({
  [`&.${classes.sectionHeader}`]: {
    padding: '1em',
  },
});

export default function ListHeader(props: { title: string; actions?: React.ReactElement[] }) {
  const actions = props.actions || [];
  return (
    <StyledGrid
      container
      alignItems="flex-start"
      justifyContent="space-between"
      className={classes.sectionHeader}
    >
      {props.title && (
        <Grid>
          <Box p={0.5}>
            <Typography variant="h1">{props.title}</Typography>
          </Box>
        </Grid>
      )}
      {actions && actions.map((action, i: number) => <Grid key={i}>{action}</Grid>)}
    </StyledGrid>
  );
}
