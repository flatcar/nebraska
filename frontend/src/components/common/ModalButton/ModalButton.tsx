import AddIcon from '@mui/icons-material/Add';
import { IconButton } from '@mui/material';
import PropTypes from 'prop-types';
import React from 'react';
import { useTranslation } from 'react-i18next';

import ApplicationEdit from '../../Applications/ApplicationEdit';
import ChannelEdit from '../../Channels/ChannelEdit';
import GroupEditDialog from '../../Groups/GroupEditDialog/GroupEditDialog';
import PackageEditDialog, {
  EditDialogProps as PackageEditDialogProps,
} from '../../Packages/EditDialog';

function ModalButton(props: { data: object; modalToOpen: string; icon?: string }) {
  const [showModal, setShowModal] = React.useState(false);
  const { t } = useTranslation();

  function close() {
    setShowModal(false);
  }

  function open() {
    setShowModal(true);
  }

  const options = {
    create: true,
    show: showModal,
    data: props.data,
    onHide: close,
  };

  let modal = null;
  switch (props.modalToOpen) {
    case 'AddApplicationModal':
      modal = <ApplicationEdit {...options} />;
      break;
    case 'AddGroupModal':
      modal = <GroupEditDialog {...options} />;
      break;
    case 'AddChannelModal':
      modal = <ChannelEdit {...options} />;
      break;
    case 'AddPackageModal':
      modal = <PackageEditDialog {...(options as PackageEditDialogProps)} />;
      break;
  }

  // @todo: verify whether aria-label should be more specific (in which
  // case it should be set from the caller).
  return (
    <div>
      <IconButton
        size="small"
        aria-label={t('frequent|add_lower')}
        onClick={open}
        data-testid="modal-button"
      >
        <AddIcon fontSize="large" sx={{ color: theme => theme.palette.titleColor }} />
      </IconButton>
      {modal}
    </div>
  );
}

ModalButton.propTypes = {
  modalToOpen: PropTypes.string.isRequired,
  data: PropTypes.object,
};

export default ModalButton;
