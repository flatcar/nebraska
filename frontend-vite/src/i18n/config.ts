import i18next from 'i18next';
import LanguageDetector from 'i18next-browser-languagedetector';
import { initReactI18next } from 'react-i18next';

import activity from './locales/en/activity.json';
import applications from './locales/en/applications.json';
import channels from './locales/en/channels.json';
import common from './locales/en/common.json';
import frequent from './locales/en/frequent.json';
import groups from './locales/en/groups.json';
import header from './locales/en/header.json';
import instances from './locales/en/instances.json';
import layouts from './locales/en/layouts.json';
import missing from './locales/en/missing.json';
import packages from './locales/en/packages.json';

i18next
  // detect user language https://github.com/i18next/i18next-browser-languageDetector
  .use(LanguageDetector)
  .use(initReactI18next)
  // i18next options: https://www.i18next.com/overview/configuration-options
  .init({
    resources: {
      en: {
        missing,
        activity,
        applications,
        channels,
        common,
        frequent,
        groups,
        header,
        instances,
        layouts,
        packages,
      },
    },
    debug: import.meta.env.NODE_ENV === 'development',
    fallbackLng: 'en',
    supportedLngs: ['en'],
    // nonExplicitSupportedLngs: true,
    interpolation: {
      escapeValue: false, // not needed for react as it escapes by default
      format: function (value, format, lng) {
        // https://www.i18next.com/translation-function/formatting
        if (format === 'number') return new Intl.NumberFormat(lng).format(value);
        if (format === 'date')
          return new Intl.DateTimeFormat(lng, {
            day: '2-digit',
            month: '2-digit',
            year: 'numeric',
          }).format(value);

        return value;
      },
    },
    returnEmptyString: false,
    // https://react.i18next.com/latest/i18next-instance
    // https://www.i18next.com/overview/configuration-options
    react: {
      useSuspense: false, // not needed unless loading from public/locales
      //   bindI18nStore: 'added'
    },
    nsSeparator: '|',
    keySeparator: false,
  });

export default i18next;
