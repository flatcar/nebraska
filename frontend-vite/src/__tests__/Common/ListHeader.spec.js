import '../../i18n/config.ts';
import { StyledEngineProvider, ThemeProvider } from '@mui/material/styles';
import { render } from '@testing-library/react';
import ListHeader from '../../components/common/ListHeader';
import ModalButton from '../../components/common/ModalButton';
import themes from '../../lib/themes';

describe('List Header', () => {
  const minProps = {
    title: 'Applications',
    actions: [<ModalButton modalToOpen="AddApplicationModal" data={{ applications: [] }} />],
  };
  it('should render correct List Header title', () => {
    const { getByText } = render(
      <StyledEngineProvider injectFirst>
        <ThemeProvider theme={themes['light']}>
          <ListHeader title={minProps.title} />
        </ThemeProvider>
      </StyledEngineProvider>
    );
    expect(getByText(minProps.title)).toBeInTheDocument();
  });
  it('should render correct List Header actions', () => {
    const { asFragment } = render(
      <StyledEngineProvider injectFirst>
        <ThemeProvider theme={themes['light']}>
          <ListHeader actions={minProps.actions} />
        </ThemeProvider>
      </StyledEngineProvider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
