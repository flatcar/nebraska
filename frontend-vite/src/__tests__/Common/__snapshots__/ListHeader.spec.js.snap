// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Header should render correct List Header actions 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid-root MuiGrid-container makeStyles-sectionHeader-2 css-eei0dj-MuiGrid-root"
  >
    <div
      class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
    >
      <div>
        <button
          aria-label="add"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
          data-testid="modal-button"
          tabindex="0"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge makeStyles-root-3 css-tzssek-MuiSvgIcon-root"
            data-testid="AddIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6z"
            />
          </svg>
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
